enzyme/Enzyme/MLIR/Passes/RemoveUnusedEnzymeOps.cpp

```diff
@@ -0,0 +1,126 @@
+//===- RemoveUnusedEnzymeOps.cpp - Remove unnecessary or unused gradient and cache ops
+//------------------ //
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements a pass to lower custom ops generated by the Enzyme AD
+// procedure to the MemRef dialect.
+//===----------------------------------------------------------------------===//
+
+#include "Dialect/Dialect.h"
+#include "Dialect/Ops.h"
+#include "PassDetails.h"
+#include "Passes/Passes.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/ControlFlow/IR/ControlFlowOps.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+#include "mlir/Rewrite/PatternApplicator.h"
+
+#include "llvm/Support/raw_ostream.h"
+#include "mlir/IR/Dominance.h"
+
+using namespace mlir;
+using namespace enzyme;
+using llvm::errs;
+namespace {
+
+// TODO: Expand to region branches??
+bool reachable(Operation * a, Operation * b){
```

Comment:
Is this function meaningfully different than `DominanceInfo`?

Reply:
I tried to find something and came up short, so I think so.